docs/compact_ranges.md

```diff
@@ -0,0 +1,68 @@
+
+## Definition
+
+For a range `[L, R)` of leaves in a Merkle tree, a Compact Range is the minimal set of nodes that “cover” these leaves (and only). For example, in the picture below the range `[2, 9)` is covered by nodes `[1.1, 2.1, 8]`, and the range `[12, 16)` is covered by a single node `2.3`.
+
+![compact_ranges](https://user-images.githubusercontent.com/3757441/148389766-9ab85a79-d40d-4f0c-a83f-d348de584e8d.png)
+
+A compact range always consists of nodes that are “final”, i.e. the hashes of these nodes never change as the tree grows. For example, range `[0, 21)` in the picture above is covered by nodes `[4.0, 2.4, 20]`, not just node `5.0`. Nodes `5.0` and `3.2` are “ephemeral” for the tree of this size, and will change when new leaves are appended to the tree until the tree size crosses their corresponding perfect subtree right borders. For simplicity, when we talk about compact ranges, we can assume that the “ephemeral” nodes don’t exist.
+
+Compact ranges have many useful properties some of which are elaborated in sections below. The basic property is that the number of nodes in a compact range `[L, R)` is `O(log(R-L))`, or `O(log N)` more generally. A compact range is always unique, and its shape is determined using a few bitwise operations on `L` and `R`.
+
+## Merging Compact Ranges
+
+The core property that makes compact ranges widely usable is that they are “mergeable”. Two compact ranges, `[L, M)` and `[M, R)`, can be efficiently merged into an `[L, R)` range. Consider the picture below for an intuition on how it works.
+
+![compact_ranges_merge](https://user-images.githubusercontent.com/3757441/148390790-d0fdd729-9846-4382-a681-8d0fe049a06b.png)
+
+Given 2 compact ranges, `[2, 9)` and `[9, 16)`, each represented by a set of node hashes (3 green and 3 cyan nodes correspondingly), we “merge” 2 sibling nodes by computing their parent’s hash any time they are both present in the set of nodes. This process repeats until there are no siblings in the set. As a result, we get hashes of nodes `[1.1, 2.1, 3.1]` which, as it turns out, represent a compact range of `[2, 16)`.
+
+Note that, when merging 2 compact ranges, the set of “new” nodes (marked in yellow) that are generated as a side effect, forms a partial path towards the root of the tree. It can be proven that this is always the case, which is a convenient property for implementations.
+
+Merging two compact ranges can be implemented in `O(log(R-L))`, or more generally `O(log N)` time. This follows from the observation in the paragraph above, and the fact that the size of the resulting `[L, R)` compact range is limited by the same estimate.
+
+## Merkle Tree Proofs
+
+A compact range `[L, R)`, if represented by the cryptographic hashes of the corresponding nodes, can be considered as a commitment to the containment of the range. The ability to merge compact ranges is effectively the ability to merge commitments.
```

Comment:
Does `containment` have a special/academic meaning here?
I've not seen it used like this before, could it be re-worded e.g. "... considered a commitment to the contents of the leaves in this range"?

Reply:
Done.